### PR TITLE
VSCode Zip Support Tooling

### PR DESCRIPTION
**Pull request by elmpp**
**What's the problem this PR addresses?**
Will be an ongoing PR for implementation of [1157](https://github.com/yarnpkg/berry/issues/1157#issuecomment-608663584)

At present:
 - integrates typescript server patching to suffix fs paths with the required `zip://` scheme.
 - procedure script accepts path to and will build working copy of VSCode
...

**How did you fix it?**
 - @yarnpkg/pnpify::generateSdk inlines the TS server patching when pnpifying
...

**To run/preview**
`./scripts/vscode-zip-test-procedure.sh ../path-to-vscode-checkout`

### CODE REVIEW

**Comment by arcanis** [comment]
@elmpp I'm trying it out locally, but the `--install-extension` flag from the VSCode doesn't seem to work. Do you have an idea if something special is needed, by any chance? 🤔

**Comment by elmpp** [comment]
@arcanis I did find one time that there was some user state overriding the extension flag. I had to check out a fresh working copy iirc.

**Comment by arcanis** [comment]
Since VSCode is gonna merge your PR, I think we can go ahead and prepare this one for merge! Do you still have some extra bandwidth for this, @elmpp? 😃

**Comment by elmpp** [comment]
yup, should  have some time now. I've left any comments on here sit whilst VsCode folk do their investigating.  
I can probably cobble together a windows test env

**Comment by jeysal** [comment]
I'm wondering how we can make this work with vim as well. Vim's default ZIP handling can open a `zipfile:/path/to.zip::/path/inside/zip`. Currently, vim is sent to `.yarn/cache/dep.zip/node_modules/dep/file.js`, which it fails to open.
Maybe something like https://github.com/lambdalisue/vim-protocol/blob/7e1cb5c4acdc8da175cc55131f316a912f998b16/autoload/protocol/zip.vim (found through random google, looks old and I never used) will allow it to handle this. Otherwise we'd need to look into making pnpify sdk support vim zipfile paths.

**Comment by elmpp** [comment]
Really couldn't find a way to get a meaningful integration test in - [this was](https://hastebin.com/unodehazic
) an abandoned approach

Tests:
 - Adds a very naive unit test as a touching point for those with more cross-platform experience to add to 🙏 
 - Have manually tested behaviour with fresh extension build and latest VSCode on OSX

File: packages/yarnpkg-pnpify/sources/generateSdk.ts [hunk 1]
```diff
@@ -178,6 +183,62 @@ export const generateTypescriptLanguageServerWrapper = async (pnpApi: PnpApi, ta
...
+                throw e;
+            }
+            return originalSend.call(this, processed);
+            // return originalSend.call(this, msg);
+        }
+    });
+
+    function addZipPrefix(str) {
+        if(str.match(/\.zip\\//) && !str.match(/^zip:\\/\\//)) {
+            return \`zip:\${str}\`;
```
[the remark this answers is not in the document]
Comment (elmpp) [reply packages/yarnpkg-pnpify/sources/generateSdk.ts:241]: Hi @JacquesBergou 

I'm without a windows env and running bit blind. Would you be able to add a couple of failing cases into [this testMap](https://github.com/yarnpkg/berry/pull/1165/files#diff-af3d43e4f75eafb5f6204796e7c7a001R100) and i can then update the patching code?

**Comment by arcanis** [comment]
I've refactored the test to directly use `tsserver.js`, otherwise there was too high a chance that we would accidentally cause a regression. I've tried it with the VSCode Nightly build and it seems to work super well! So well that I actually initially thought I forgot something and the files were unplugged 😄

I'll wait to make sure that all the tests pass, but it should be merged very soon. Thanks you all for your work, particularly @elmpp and @cspotcode!

File: packages/yarnpkg-pnpify/sources/generateSdk.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (arcanis) [reply packages/yarnpkg-pnpify/sources/generateSdk.ts]: @JacquesBergou From a quick test I'm not sure I see why the leading forward slash is required on Windows; if that's the case can you please open a PR to fix it and detail how to reproduce the crash on Windows? Thanks!

**Pull request merged**